node/src/exchange/client.rs

```diff
@@ -56,63 +96,697 @@ impl ExchangeClientHandler {
+        assert!(matches!(
+            rx.await,
+            Ok(Err(P2pError::Exchange(ExchangeError::InvalidRequest)))
+        ));
+
+        // Head with zero amount
+        let (tx, rx) = oneshot::channel();
+        handler.on_send_request(&mut mock_req, HeaderRequest::with_origin(0, 0), tx);
+        assert!(matches!(
+            rx.await,
+            Ok(Err(P2pError::Exchange(ExchangeError::InvalidRequest)))
+        ));
+
+        // Head with more than one amount
+        let (tx, rx) = oneshot::channel();
+        handler.on_send_request(&mut mock_req, HeaderRequest::with_origin(0, 2), tx);
+        assert!(matches!(
+            rx.await,
+            Ok(Err(P2pError::Exchange(ExchangeError::InvalidRequest)))
+        ));
+
+        // Invalid hash
+        let (tx, rx) = oneshot::channel();
+        handler.on_send_request(
+            &mut mock_req,
+            HeaderRequest {
+                data: Some(Data::Hash(b"12".to_vec())),
+                amount: 1,
+            },
+            tx,
+        );
+        assert!(matches!(
+            rx.await,
+            Ok(Err(P2pError::Exchange(ExchangeError::InvalidRequest)))
+        ));
+
+        // Valid hash with more than one amount
+        let (tx, rx) = oneshot::channel();
+        handler.on_send_request(
+            &mut mock_req,
+            HeaderRequest {
+                data: Some(Data::Hash([0xff; HASH_SIZE].to_vec())),
+                amount: 2,
+            },
+            tx,
+        );
+        assert!(matches!(
+            rx.await,
+            Ok(Err(P2pError::Exchange(ExchangeError::InvalidRequest)))
+        ));
+
+        // No data
+        let (tx, rx) = oneshot::channel();
+        handler.on_send_request(
+            &mut mock_req,
+            HeaderRequest {
+                data: None,
+                amount: 2,
+            },
+            tx,
+        );
+        assert!(matches!(
+            rx.await,
+            Ok(Err(P2pError::Exchange(ExchangeError::InvalidRequest)))
+        ));
+    }
+
+    /// Expects the highest heigh that was reported by at least 2 peers
+    #[tokio::test]
+    async fn head_best() {
+        let peer_tracker = peer_tracker_with_n_peers(15);
+        let mut mock_req = MockReq::new();
+        let mut handler = ExchangeClientHandler::<MockReq>::new(peer_tracker);
+
+        let (tx, rx) = oneshot::channel();
+
+        handler.on_send_request(&mut mock_req, HeaderRequest::with_origin(0, 1), tx);
+
+        let expected = gen_header_response(5);
+        let expected_header = expected.to_extended_header().unwrap();
+
+        mock_req.send_n_responses(&mut handler, 1, vec![gen_header_response(3)]);
+        mock_req.send_n_responses(&mut handler, 2, vec![gen_header_response(4)]);
+        mock_req.send_n_responses(&mut handler, 1, vec![gen_header_response(5)]);
+        mock_req.send_n_responses(&mut handler, 2, vec![expected]);
```

Comment:
not sure if that matters, but we actually send height 5 three times here

Reply:
It was on purpose. The first one has different hash than the `expected`. Algorithm should choose the `expected`.

Reply:
:100: 
Can we have a comment so I won't miss that again?

```suggestion
        // this header also has height = 5 but has different hash
        mock_req.send_n_responses(&mut handler, 1, vec![gen_header_response(5)]);
        mock_req.send_n_responses(&mut handler, 2, vec![expected]);
```